mrp_bom_tracking/models/mrp_bom.py

```diff
@@ -0,0 +1,96 @@
+                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
+                    )
+                bom_line_ids[bom.id] = bom.bom_line_ids
+        res = super().write(values)
+        if "bom_line_ids" in values:
+            for bom in self:
+                new_lines = bom.bom_line_ids - bom_line_ids[bom.id]
+                if new_lines:
+                    message_body = template_env._render_template(
+                        "mrp_bom_tracking.track_bom_template",
+                        values={"lines": new_lines, "mode": "New"},
+                    )
+                    bom.message_post(
+                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
+                    )
+        return res
+
+
+class MrpBomLine(models.Model):
+    _inherit = "mrp.bom.line"
+
+    def write(self, values):
+        template_env = self.env["ir.ui.view"]
+        if "product_id" in values:
+            for bom in self.mapped("bom_id"):
+                # Bind bom variable for the current iteration to the lambda
+                lines = self.filtered(lambda x, bom=bom: x.bom_id == bom)
+                product_id = values.get("product_id")
+                if product_id:
+                    product_id = self.env["product.product"].browse(product_id)
+                product_id = product_id or lines.product_id
+                if lines:
+                    message_body = template_env._render_template(
+                        "mrp_bom_tracking.track_bom_template_2",
+                        values={"lines": lines, "product_id": product_id},
+                    )
+                    bom.message_post(
+                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
+                    )
+        elif "product_qty" in values or "product_uom_id" in values:
+            for bom in self.mapped("bom_id"):
+                # Bind bom variable for the current iteration to the lambda
+                lines = self.filtered(lambda x, bom=bom: x.bom_id == bom)
+                if lines:
+                    product_qty = values.get("product_qty") or lines.product_qty
+                    product_uom_id = values.get("product_uom_id")
+                    if product_uom_id:
+                        product_uom_id = self.env["uom.uom"].browse(product_uom_id)
+                    product_uom_id = product_uom_id or lines.product_uom_id
+
+                    message_body = template_env._render_template(
+                        "mrp_bom_tracking.track_bom_line_template",
+                        values={
+                            "lines": lines,
+                            "product_qty": product_qty,
+                            "product_uom_id": product_uom_id,
+                        },
+                    )
+                    bom.message_post(
+                        body=message_body, subtype_id=self.env.ref("mail.mt_note").id
+                    )
```

Comment:
```suggestion
        elif "product_qty" in values or "product_uom_id" in values:
            for bom_line in self:
                bom = bom_line.bom_id
                product_qty = values.get("product_qty") or bom_line.product_qty
                product_uom_id = values.get("product_uom_id")
                if product_uom_id:
                    product_uom = self.env["uom.uom"].browse(product_uom_id)
                product_uom = product_uom or bom_line.product_uom_id
                message_body = template_env._render_template(
                    "mrp_bom_tracking.track_bom_line_template",
                    values={
                        "lines": bom_line,
                        "product_qty": product_qty,
                        "product_uom_id": product_uom,
                    },
                )
                bom.message_post(
                    body=message_body, subtype_id=self.env.ref("mail.mt_note").id
                )
```
Same as above comment.